File: BaeKit/Sources/BaeKit/DisconnectSyncFlow.swift
import Foundation
import Observation
import os.log

private let logger = Logger.bae("DisconnectSyncFlow")

/// Confirmation-and-execution model for disconnecting the cloud provider from
/// this device, shared by the macOS and iOS settings screens. Built with
/// injected closures so each platform supplies its live bridge/keychain calls,
/// its localized message text (the base sentence differs — iOS notes that
/// reconnecting needs another device because it has no provider-setup flow,
/// macOS omits that because it does), and its error-string templates (tests and
/// the shared package can't resolve an app catalog's localized strings, so the
/// app resolving them owns them). The at-risk sentence is owned the same way:
/// bae-core supplies the count, the app pluralizes it for its locale.
///
/// Two ordering invariants the flow enforces:
/// - the disconnect suspends through the bridge while bae-core stops the sync
///   loop on its owned runtime;
/// - the restore code is deleted from the iCloud keychain only after a
///   successful disconnect — the code embeds the cloud-home connection that
///   disconnect invalidates, so a failed disconnect must leave it in place.
@MainActor
@Observable
public final class DisconnectSyncFlow {
    /// How many releases live only in the cloud, from bae-core. `0` means nothing
    /// is at risk.
    private let cloudOnlyReleaseCount: @Sendable () async throws -> UInt64
    /// Renders that count as the localized at-risk sentence, with the app's own
    /// plural rules (`core.sync.cloud_only_releases`). Only called for counts > 0.
    private let atRiskMessage: (UInt64) -> String
    /// Disconnects the provider through bae-core's owned runtime.
    private let disconnect: @Sendable () async throws -> Void
    /// Removes this library's restore code from the iCloud keychain. Throws
    /// when the keychain refuses; the disconnect itself has already landed by
    /// then, so that is its own message rather than a failed disconnect.
    private let deleteRestoreCode: () throws -> Void
    /// The platform's base confirmation sentence, resolved against the app's
    /// localized catalog.
    private let baseMessage: () -> String
    /// Builds the inline error shown when the at-risk check itself fails, from
    /// the error's description.
    private let warningCheckFailedMessage: (String) -> String
    /// Builds the inline error shown when the disconnect fails, from the error's
    /// description.
    private let disconnectFailedMessage: (String) -> String
    /// Builds the inline error shown when the disconnect succeeded but the
    /// keychain would not drop the restore code.
    private let restoreCodeDeleteFailedMessage: (String) -> String

    /// Whether the confirmation dialog is shown.
    public var showConfirm = false
    /// The at-risk sentence to append to the confirmation body, or `nil` when
    /// nothing is at risk (or the check failed).
    public var extraWarning: String?
    /// Inline error line for the Sync section, or `nil`.
    public var error: String?

    @ObservationIgnored
    private var warningTask: Task<Void, Never>?

    public init(
        cloudOnlyReleaseCount: @escaping @Sendable () async throws -> UInt64,
        atRiskMessage: @escaping (UInt64) -> String,
        disconnect: @escaping @Sendable () async throws -> Void,
        deleteRestoreCode: @escaping () throws -> Void,
        baseMessage: @escaping () -> String,
        warningCheckFailedMessage: @escaping (String) -> String,
        disconnectFailedMessage: @escaping (String) -> String,
        restoreCodeDeleteFailedMessage: @escaping (String) -> String
    ) {
        self.cloudOnlyReleaseCount = cloudOnlyReleaseCount
        self.atRiskMessage = atRiskMessage
        self.disconnect = disconnect
        self.deleteRestoreCode = deleteRestoreCode
        self.baseMessage = baseMessage
        self.warningCheckFailedMessage = warningCheckFailedMessage
        self.disconnectFailedMessage = disconnectFailedMessage
        self.restoreCodeDeleteFailedMessage = restoreCodeDeleteFailedMessage
    }

    /// Confirmation body: the platform's base sentence and — when releases live
    /// only in the cloud — the localized at-risk sentence appended to it.
    public var message: String {
        guard let extraWarning else { return baseMessage() }
        return "\(baseMessage()) \(extraWarning)"
    }

    /// Query the at-risk warning, then open the confirmation. If the query
    /// itself fails, surface the error inline (so the user sees the data-loss
    /// check didn't run) and still open the confirmation so they can proceed
    /// or cancel.
    public func promptDisconnect() {
        error = nil
        warningTask?.cancel()
        warningTask = Task {
            do {
                let count = try await cloudOnlyReleaseCount()
                extraWarning = count > 0 ? atRiskMessage(count) : nil
            }
            catch is CancellationError {
                return
            }
            catch {
                logger.error(
                    "Failed to compute disconnect warning: \(error.localizedDescription)"
                )
                // No line means nothing to say — a cancellation. Don't invent one.
                self.error = error.displayLine.map(warningCheckFailedMessage)
                extraWarning = nil
            }
            showConfirm = true
        }
    }

    /// Cancel an in-flight warning query when the view disappears.
    public func cancelWarningTask() {
        warningTask?.cancel()
    }

    /// Disconnect, then — only on success — delete the restore code. A failed
    /// disconnect leaves the code untouched and shows its error inline.
    ///
    /// A keychain that refuses the delete is reported separately, because it is
    /// not a failed disconnect: the provider is gone, and what is left is a
    /// restore code still naming the cloud home the disconnect invalidated.
    /// Saying "failed to disconnect" there would point the user at the wrong
    /// thing, and saying nothing leaves a code they think was removed.
    public func confirm() async {
        do {
            try await disconnect()
        }
        catch {
            logger.error("Failed to disconnect: \(error.localizedDescription)")
            self.error = error.displayLine.map(disconnectFailedMessage)
            return
        }
        error = nil
        do {
            try deleteRestoreCode()
        }
        catch {
            logger.error(
                "Disconnected, but failed to delete the restore code: \(error.localizedDescription)"
            )
            self.error = error.displayLine.map(restoreCodeDeleteFailedMessage)
        }
    }
}

File: bae-android/app/src/main/java/fm/bae/app/ui/settings/DisconnectSyncFlow.kt
package fm.bae.app.ui.settings

import fm.bae.app.BaeLogger
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val TAG = "bae.DisconnectSyncFlow"
private val logger = BaeLogger(TAG)

/**
 * Confirmation-dialog visibility and inline-error state for disconnecting the
 * cloud provider from this device.
 */
internal data class DisconnectSyncState(
    /** Whether the confirmation dialog is shown. */
    val confirming: Boolean = false,
    /**
     * The localized at-risk sentence to append to the confirmation body, or null
     * when nothing is at risk (or the check itself failed).
     */
    val extraWarning: String? = null,
    /** Inline error line in the Sync section, or null. */
    val error: String? = null,
)

/**
 * The confirmation body: a localized [base] sentence with the at-risk sentence
 * appended after a single space when present. bae-core supplies the count; the
 * sentence itself is resolved here from `core.sync.cloud_only_releases`, with
 * this locale's plural rules.
 */
internal fun disconnectConfirmMessage(
    base: String,
    extraWarning: String?,
): String = if (extraWarning.isNullOrEmpty()) base else "$base $extraWarning"

/**
 * The localized lines the disconnect flow shows: the at-risk warning for a
 * cloud-only release count, and the failure lines for a warning check or a
 * disconnect that errored. The screen supplies live resource-backed formatters;
 * tests supply stubs.
 */
internal data class DisconnectStrings(
    val atRiskLine: (ULong) -> String,
    /** Null when core says the failure has no line to show — a cancellation. */
    val warningFailedLine: (Throwable) -> String?,
    /** Null when core says the failure has no line to show — a cancellation. */
    val disconnectFailedLine: (Throwable) -> String?,
)

/**
 * Drives the disconnect confirmation and execution with injected bridge calls so
 * the screen supplies the live handle and tests supply stubs.
 *
 * Two invariants:
 * - the disconnect runs off the main thread, because bae-core joins the
 *   sync-loop thread and can block for the remainder of an in-flight cycle;
 * - a failed at-risk check still opens the confirmation, surfacing the failure
 *   inline so the user sees the data-loss check didn't run yet can still proceed.
 */
internal class DisconnectSyncFlow(
    private val scope: CoroutineScope,
    private val cloudOnlyReleaseCount: suspend () -> ULong,
    private val disconnect: () -> Unit,
    private val strings: DisconnectStrings,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
) {
    private val _state = MutableStateFlow(DisconnectSyncState())
    val state: StateFlow<DisconnectSyncState> = _state.asStateFlow()

    private var warningJob: Job? = null

    /**
     * Query the at-risk warning, then open the confirmation. On query failure,
     * surface the error inline and still open the confirmation so the user can
     * proceed or cancel. A superseding call cancels the prior query.
     */
    fun promptDisconnect() {
        _state.update { it.copy(error = null) }
        warningJob?.cancel()
        warningJob =
            scope.launch {
                _state.value =
                    try {
                        val count = cloudOnlyReleaseCount()
                        DisconnectSyncState(
                            confirming = true,
                            extraWarning = if (count > 0uL) strings.atRiskLine(count) else null,
                        )
                    } catch (e: CancellationException) {
                        throw e
                    } catch (e: Throwable) {
                        logger.error("Failed to compute disconnect warning", e)
                        DisconnectSyncState(confirming = true, error = strings.warningFailedLine(e))
                    }
            }
    }

    fun dismissConfirm() {
        _state.update { it.copy(confirming = false) }
    }

    /**
     * Disconnect off the main thread. On success the sync config is cleared by
     * core and the whole connected section falls away via config invalidation;
     * on failure the error shows inline and nothing durable changed.
     */
    suspend fun confirm() {
        _state.value =
            try {
                withContext(ioDispatcher) { disconnect() }
                DisconnectSyncState()
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                logger.error("Failed to disconnect cloud provider", e)
                DisconnectSyncState(error = strings.disconnectFailedLine(e))
            }
    }
}
